metricbeat/module/windows/perfmon/pdh_query_windows.go

```diff
@@ -0,0 +1,211 @@
+	instanceName string
+}
+
+// Query contains the pdh.
+type Query struct {
+	handle   PdhQueryHandle
+	counters map[string]*Counter
+}
+
+// CounterValue contains the performance counter values.
+type CounterValue struct {
+	Instance    string
+	Measurement interface{}
+	Err         error
+}
+
+// Open creates a new query.
+func (q *Query) Open() error {
+	h, err := PdhOpenQuery("", 0)
+	if err != nil {
+		return err
+	}
+	q.handle = h
+	q.counters = make(map[string]*Counter)
+	return nil
+}
+
+// AddCounter adds the specified counter to the query.
+func (q *Query) AddCounter(counterPath string, counter CounterConfig, wildcard bool) error {
+	if _, found := q.counters[counterPath]; found {
+		return errors.Errorf("Counter %s has been already added", counterPath)
+	}
+	var err error
+	var instanceName string
+	// Extract the instance name from the counterPath.
+	if wildcard || (!wildcard && counter.InstanceName == "") {
+		instanceName, err = matchInstanceName(counterPath)
+		if err != nil {
+			return err
+		}
+	} else {
+		instanceName = counter.InstanceName
+	}
+	h, err := PdhAddCounter(q.handle, counterPath, 0)
+	if err != nil {
+		return err
+	}
+
+	q.counters[counterPath] = &Counter{
+		handle:       h,
+		instanceName: instanceName,
+		format:       getFormat(counter.Format),
+	}
+	return nil
+}
+
+// CollectData collects the value for all counters in the query.
+func (q *Query) CollectData() error {
+	return PdhCollectQueryData(q.handle)
+}
+
+// GetFormattedCounterValues returns an array of formatted values for a query.
+func (q *Query) GetFormattedCounterValues() (map[string][]CounterValue, error) {
+	rtn := make(map[string][]CounterValue, len(q.counters))
+
+	for path, counter := range q.counters {
+		_, value, err := PdhGetFormattedCounterValue(counter.handle, counter.format|PdhFmtNoCap100)
+		if err != nil {
+			rtn[path] = append(rtn[path], CounterValue{Err: err})
+			continue
+		}
+
+		switch counter.format {
+		case PdhFmtDouble:
+			rtn[path] = append(rtn[path], CounterValue{Measurement: *(*float64)(unsafe.Pointer(&value.LongValue)), Instance: counter.instanceName})
+		case PdhFmtLarge:
+			rtn[path] = append(rtn[path], CounterValue{Measurement: *(*int64)(unsafe.Pointer(&value.LongValue)), Instance: counter.instanceName})
+
+		}
+	}
+
+	return rtn, nil
+}
+
+// ExpandWildCardPath  examines local computer and returns those counter paths that match the given counter path which contains wildcard characters.
+func (q *Query) ExpandWildCardPath(wildCardPath string) ([]string, error) {
+	if wildCardPath == "" {
+		return nil, errors.New("no wildcardpath given")
+	}
+
+	expdPaths, err := PdhExpandWildCardPath(wildCardPath)
+	if err != nil {
+		return nil, err
+	}
+	return UTF16ToStringArray(expdPaths), nil
+
+}
+
+// Close closes the query and all of its counters.
+func (q *Query) Close() error {
+	return PdhCloseQuery(q.handle)
+}
+
+// Error returns a more explicit error message.
+func (e PdhErrno) Error() string {
```

Comment:
Do we have some tests for this method somehwere? Pretty hard to follow what the end outcome is without a test.

Reply:
BTW I wonder if this method belongs here or rather in the phd_windows file.

Reply:
indeed, there is already a test in pdh_windows_test file, and I have moved this func to the pdh_windows file